drivers/serial/uart_rpi_pico_pio.c

```diff
@@ -0,0 +1,172 @@
+#define SIDESET_BIT_COUNT 2
+
+struct pio_uart_config {
+	PIO pio;
+	uint32_t baudrate;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_tx, 0, 3,
+		/* .wrap_target */
+	0x9fa0, /*  0: pull   block           side 1 [7]  */
+	0xf727, /*  1: set    x, 7            side 0 [7]  */
+	0x6001, /*  2: out    pins, 1                     */
+	0x0642, /*  3: jmp    x--, 2                 [6]  */
+		/* .wrap */
+);
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_rx, 0, 8,
+		/*  .wrap_target */
+	0x2020, /*  0: wait   0 pin, 0                    */
+	0xea27, /*  1: set    x, 7                   [10] */
+	0x4001, /*  2: in     pins, 1                     */
+	0x0642, /*  3: jmp    x--, 2                 [6]  */
+	0x00c8, /*  4: jmp    pin, 8                      */
+	0xc014, /*  5: irq    nowait 4 rel                */
+	0x20a0, /*  6: wait   1 pin, 0                    */
+	0x0000, /*  7: jmp    0                           */
+	0x8020, /*  8: push   block                       */
+		/*  .wrap */
+);
+
+static void pio_uart_tx_init(PIO pio, uint32_t pin_tx, float div)
+{
+	uint32_t offset;
+	pio_sm_config sm_config;
+
+	offset = pio_add_program(pio, &uart_tx_program);
+	sm_config = pio_get_default_sm_config();
+
+	sm_config_set_sideset(&sm_config, SIDESET_BIT_COUNT, true, false);
+	sm_config_set_out_shift(&sm_config, true, false, 0);
+	sm_config_set_out_pins(&sm_config, pin_tx, 1);
+	sm_config_set_sideset_pins(&sm_config, pin_tx);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, div);
+	sm_config_set_wrap(&sm_config,
+			   offset + RPI_PICO_PIO_GET_WRAP_TARGET(uart_tx),
+			   offset + RPI_PICO_PIO_GET_WRAP(uart_tx));
+
+	pio_sm_set_pins_with_mask(pio, TX_STATE_MACHINE, BIT(pin_tx), BIT(pin_tx));
+	pio_sm_set_pindirs_with_mask(pio, TX_STATE_MACHINE, BIT(pin_tx), BIT(pin_tx));
+	pio_sm_init(pio, TX_STATE_MACHINE, offset, &sm_config);
+	pio_sm_set_enabled(pio, TX_STATE_MACHINE, true);
+}
+
+static void pio_uart_rx_init(PIO pio, uint32_t pin, float div)
+{
+	uint32_t offset;
+	pio_sm_config sm_config;
+
+	offset = pio_add_program(pio, &uart_rx_program);
+	sm_config = pio_get_default_sm_config();
+
+	pio_sm_set_consecutive_pindirs(pio, RX_STATE_MACHINE, pin, 1, false);
+	sm_config_set_in_pins(&sm_config, pin);
+	sm_config_set_jmp_pin(&sm_config, pin);
+	sm_config_set_in_shift(&sm_config, true, false, 0);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_RX);
+	sm_config_set_clkdiv(&sm_config, div);
+	sm_config_set_wrap(&sm_config,
+			   offset + RPI_PICO_PIO_GET_WRAP_TARGET(uart_rx),
+			   offset + RPI_PICO_PIO_GET_WRAP(uart_rx));
+
+	pio_sm_init(pio, RX_STATE_MACHINE, offset, &sm_config);
+	pio_sm_set_enabled(pio, RX_STATE_MACHINE, true);
+}
+
+static int pio_uart_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct pio_uart_config *config = dev->config;
+	io_rw_8 *uart_rx_fifo_msb;
+	PIO pio = config->pio;
+
+	uart_rx_fifo_msb = (io_rw_8 *)&pio->rxf[RX_STATE_MACHINE] + 3;
```

Comment:
Comment/details on what's special about the magic number 3?

Reply:
It's a byte pointer to a 32 bit register, so adding 3 gives the MSB of the register, which is denoted in the variable name. But I can document it if it's not clear enough.

Reply:
Just browsing, but yes, the 3 needs documenting. I guess there's no need to be worried about big/little endianness because the PIO is built into the soc, but if it were me I would include a comment including "BYTE_ORDER" just to be clear.